Fall 2023/Operating Systems Design/project4/rufs.c

```diff
@@ -288,10 +288,125 @@ int dir_add(struct inode dir_inode, uint16_t f_ino, const char *fname, size_t na
 	return EXIT_SUCCESS;
 }
 
+//clears data block and marks it available in data block bitmap
+void remove_data_block(int data_block_number){
+
+	//clear data block
+	struct dirent *block_of_zeroes = malloc(sizeof(BLOCK_SIZE));
+	biowrite(data_block_number, block_of_zeroes);
+	free(block_of_zeroes);
+
+	//make data block available in data bitmap
+	bitmap_t data_bitmap = get_data_bitmap(superblock);
+	unset_bitmap(data_bitmap, data_block_number);
+	update_data_bitmap(data_bitmap, TRUE, superblock);
+
+}
+
+//clears an inode (setting mem to all 0 will make it invalid) and removes it from inode bitmap
+void remove_inode(int inode_number){
+
+	//clear data in inode
+	struct inode zero;
+	memset(&zero, 0, sizeof(struct inode));
+	writei(inode_number, &zero);
+
+	//mark cleared inode as available in inode bitmap
+	bitmap_t inode_bitmap = get_inode_bitmap(superblock);
+	unset_bitmap(inode_bitmap, inode_bitmap);
+	update_inode_bitmap(inode_bitmap, TRUE, superblock);
+}
+
+void remove_this_file(struct inode inode_of_file_to_remove){
+
+	struct dirent *block_of_mem = malloc(sizeof(BLOCK_SIZE));
+
+	//clear any allocated blocks pointed to directly
+	for(int direct_pointer_index = 0; direct_pointer_index < 16; direct_pointer_index ++){
+		if(inode_of_file_to_remove.direct_ptr[direct_pointer_index] != 0){
+			remove_data_block(inode_of_file_to_remove.direct_ptr[direct_pointer_index]);
+		}
+	}
+
+	//clear any blocks used by indirect pointers
+	for(int indirect_pointer_index = 0; indirect_pointer_index < 8; indirect_pointer_index ++){
+		
+		//0 means it doesn't point to a block
+		if(inode_of_file_to_remove.indirect_ptr[indirect_pointer_index] == 0){
+			continue;
+		}
+
+		bio_read(inode_of_file_to_remove.indirect_ptr[indirect_pointer_index], block_of_mem);
+		
+		//free the blocks pointed to by dirents in indeirect block
+		for(int direct_pointer_index = 0; direct_pointer_index < 16; direct_pointer_index ++){
+			if(inode_of_file_to_remove.direct_ptr[direct_pointer_index] != 0){
+				remove_data_block(inode_of_file_to_remove.direct_ptr[direct_pointer_index]);
```

Comment:
Same for this line. Would it not be indirect_ptr[(in)direct_ptr_index]? I think the best thing to do in this loop is for each indirect block that is allocated, cast the current integer (0-1024) from block_of_mem and see if its nonzero, in which case it would be set to 0 and removed. So the line would be something like *(int *)(block_of_mem * sizeof(int) * i)

Reply:
> Same for this line. Would it not be indirect_ptr[(in)direct_ptr_index]? I think the best thing to do in this loop is for each indirect block that is allocated, cast the current integer (0-1024) from block_of_mem and see if its nonzero, in which case it would be set to 0 and removed. So the line would be something like *(int *)(block_of_mem * sizeof(int) * i)

I agree this is incorrect but I fixed it in an easier way